### PR TITLE
Stabilizer simulator

### PR DESCRIPTION
**Pull request by ncrubin**
Implements the "Improved Simulation of Stabilizer Circuits" from Aaronson and Gottesman.  This qvm also contains nice functions for projecting out states starting from stabilizers and returning the density matrix starting from stabilizers.

### CODE REVIEW

File: referenceqvm/api.py [hunk 1]
```diff
@@ -35,6 +35,8 @@ def QVMConnection(type_trans='wavefunction',
         elif type_trans == 'density':
             qvm = QVM_Density(gate_set=gate_set, noise_model=noise_model)
 
+        elif type_trans == 'stabilizer':
+            qvm = QVM_Stabilizer(gate_set=stabilizer_gate_set)
```
Comment (mpharrigan) [inline referenceqvm/api.py]: missing imports

File: referenceqvm/unitary_generator.py [hunk 2]
```diff
@@ -418,7 +418,9 @@ def value_get(param_obj):
         return param_obj.index
     elif isinstance(param_obj, Addr):
         return param_obj.address
-    elif isinstance(param_obj, Slot):
-        return param_obj.value()
+    # elif isinstance(param_obj, Slot):
```
Comment (mpharrigan) [inline referenceqvm/unitary_generator.py]: ?

File: referenceqvm/qvm_stabilizer.py [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+
+    V I R T U A L
+
+    M A C H I N E
+
+    S I M U L A T I N G
+
+    S T A B I L I Z E R
+
+    S T A T E S
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py:49]: lol

File: referenceqvm/qam.py [hunk 4]
```diff
@@ -91,8 +92,7 @@ def load_program(self, pyquil_program):
 
         # NOTE: all_inst is set by the subclass
         if invalid is True and self.all_inst is False:
-            raise TypeError("In QVM_Unitary, only Gates and DefGates are "
-                            "supported")
+            raise TypeError("Some gates used are not allowed in this QAM")
```
Comment (mpharrigan) [inline referenceqvm/qam.py:96]: will defgates let the user sneak through invalid programs?

File: referenceqvm/qvm_stabilizer.py [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+        unitary check QVM checks makes it annoying to allow measurements.
+        TODO: Change load_program for each qvm-subclass
+
+        :param Program pyquil_program: program to be run
+        """
+        # typecheck
+        if not isinstance(pyquil_program, Program):
+            raise TypeError("I can only generate from pyQuil programs")
+
+        if self.all_inst is None:
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py:87]: this seems out of place

File: referenceqvm/qvm_stabilizer.py [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+        # this sets that arbitrary instructions are not allowed
+        # this can probably be factored out eventually.  It is a relic of the
+        # the old days of generating unitaries of programs
+        self.all_inst = False
+        if num_qubits is None:
+            self.tableau = None
+        else:
+            self.tableau = self._n_qubit_tableau(num_qubits)
+
+    def load_program(self, pyquil_program):
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py:71]: this whole function is redundant with the superclass

File: referenceqvm/qvm_stabilizer.py [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+            raise NotImplementedError("QAM needs to be subclassed in order to "
+                                      "load program")
+
+        # create defgate dictionary
+        defined_gates = {}
+        for dg in pyquil_program.defined_gates:
+            defined_gates[dg.name] = dg.matrix
+        self.defgate_set = defined_gates
+
+        # if QVM_Unitary, check if all instructions are valid.
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py]: it's probably not `QVM_Unitary` at this point

File: referenceqvm/qvm_stabilizer.py [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+        :rtype: bool int
+        """
+        self.pre()
+        self._transition(instruction)
+        self.post()
+
+        # return HALTED (i.e. program_counter is end of program)
+        return self.program_counter == len(self.program)
+
+    def pre(self):
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py:156]: why bother with these?

File: referenceqvm/qvm_stabilizer.py [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+                pass
+            else:
+                raise TypeError("We checked for correct gate types previously" +
+                                " so the impossible has happened!")
+
+            self.program_counter += 1
+
+        elif isinstance(instruction, Jump):
+            # unconditional Jump; go directly to Label
+            self.program_counter = self.find_label(instruction.target)
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py:197]: `self.find_label` doesn't exist

File: referenceqvm/qvm_stabilizer.py (hunk 6 above)
Comment (ncrubin) [reply referenceqvm/qvm_stabilizer.py:71]: I make a note that at some point each subclass or QAM should make its own load_program.  The reason for this is that if load_program needs to check particular gate sets (like matchgates only...or something like that) each new subclass has to provide this functionality.  In the future I think I'd like to remove load_program from QAM.

Comment (mpharrigan) [reply referenceqvm/qvm_stabilizer.py:71]: then can you remove the obviously-not-appropriate-for-stabalizer-qvm parts of this implementation

File: referenceqvm/qvm_stabilizer.py [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+            # apply Gate or DefGate
+            if instruction.name == 'H':
+                self._apply_hadamard(instruction)
+            elif instruction.name == 'S':
+                self._apply_phase(instruction)
+            elif instruction.name == 'CNOT':
+                self._apply_cnot(instruction)
+            elif instruction.name == 'I':
+                pass
+            else:
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py:189]: what about X, Y, Z?

File: referenceqvm/qvm_stabilizer.py [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+
+        if classical_addresses is None:
+            classical_addresses = get_classical_addresses_from_program(pyquil_program)
+
+        results = []
+        for _ in range(trials):
+            # set up stabilizers
+            self.tableau = self._n_qubit_tableau(self.num_qubits)
+            self.kernel()
+            results.append(list(map(int, self.classical_memory[classical_addresses])))
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py:227]: `[int(b) for b in self.classical_memory[classical_addresses]]`

File: referenceqvm/qvm_stabilizer.py [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+
+        :param Program pyquil_program: a pyquil Program containing only
+                                       CNOT-H-S-MEASUREMENT operations
+        :return:
+        """
+        self.load_program(pyquil_program)
+        self.tableau = self._n_qubit_tableau(self.num_qubits)
+        self.kernel()
+        stabilizers = binary_stabilizer_to_pauli_stabilizer(self.stabilizer_tableau())
+        pauli_ops = list(map(lambda x: 0.5 * (sI(0) + x), stabilizers))
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py]: write as list comprehension

File: referenceqvm/qvm_stabilizer.py [hunk 13]
```diff
@@ -0,0 +1,479 @@
...
+
+        :param program: pyQuil program composed of stabilizer operations only
+        :return: pyquil.Wavefunction.wavefunction object.
+        """
+        self.load_program(program)
+        self.tableau = self._n_qubit_tableau(self.num_qubits)
+        self.kernel()
+        stabilizers = binary_stabilizer_to_pauli_stabilizer(self.stabilizer_tableau())
+        stabilizer_state = project_stabilized_state(stabilizers)
+        stabilizer_state = np.array(stabilizer_state.todense())  # todense() returns a matrixlib type
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py]: `stabalizer_state.toarray()` does this in one fell swoop

File: referenceqvm/qvm_stabilizer.py [hunk 14]
```diff
@@ -0,0 +1,479 @@
...
+
+        Note: this accumulates the $i$ phase for row_i * row_h  NOT row_h, row_i
+
+        :param Int h: row index 'h'
+        :param Int i: row index 'i'
+        :return: phase mod 4
+        """
+        phase_accumulator = 0
+        for j in range(self.num_qubits):
+            # self._g_update(x_{hj}, z_{hj}, x_{ij}, z_{ij})
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py]: this comment doesn't match the actual indices?

File: referenceqvm/qvm_stabilizer.py [hunk 15]
```diff
@@ -0,0 +1,479 @@
...
+    def _cnot_phase_update(self, i, c, t):
+        """
+        update r_{i} as a submethod to applying CNOT to the tableau
+
+        :param i: tableau row index
+        :param c: control qubit index
+        :param t: target qubit index
+        :return: 0/1 phase update for r_{i}
+        """
+        return self.tableau[i, -1] ^ (self.tableau[i, c] * self.tableau[i, t + self.num_qubits]) * (
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py]: this line is wrapped weirdly

File: referenceqvm/qvm_stabilizer.py [hunk 16]
```diff
@@ -0,0 +1,479 @@
...
+
+        # outcome of measurement is deterministic...need to determine sign
+        else:
+            # augment tableau with a scratch space
+            self.tableau = np.vstack((self.tableau, np.zeros((1, 2 * self.num_qubits + 1), dtype=int)))
+            for ii in range(self.num_qubits):
+                # We need to check if R(i) anticommutes with Za...which it does if x_{ia} = 1
+                if self.tableau[ii, t_qbit] == 1:  # referencing the destabilizers
+
+                    # TODO: Remove this for performance?
```
Comment (mpharrigan) [inline referenceqvm/qvm_stabilizer.py:465]: did you want to remove this?

File: referenceqvm/stabilizer_utils.py [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+
+    num_qubits = int(np.log2(state.shape[0]))
+    new_coeffs = []
+    new_indices = []
+
+    # iterate through non-zero
+    rindices, cindices = state.nonzero()
+    for ridx, cidx in zip(rindices, cindices):
+        # this is so gross looking
+        bitstring = list(map(int, np.binary_repr(ridx, width=num_qubits)))[::-1]
```
Comment (mpharrigan) [inline referenceqvm/stabilizer_utils.py]: list comprehension!

File: referenceqvm/stabilizer_utils.py [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+        new_ket, new_coefficient = compute_action(bitstring, pauli_operator, num_qubits)
+        new_indices.append(int("".join([str(x) for x in new_ket[::-1]]), 2))
```
Comment (mpharrigan) [inline referenceqvm/stabilizer_utils.py]: this seems like an odd way of going from bits to an integer. maybe make a function to make it clear what you're doing here. You can avoid converting to string if you do something like

```python
places = 2**np.arange(n)
num = np.sum(places * bits)
```

File: referenceqvm/stabilizer_utils.py [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+
+    state = csc_matrix((data, (indices, indptr)), shape=(2 ** num_qubits, 1),
+                       dtype=complex)
+
+    for generator in stabilizer_list:
+        # (I + G(i)) / 2
+        state += state_family_generator(state, generator)
+        state /= 2
+
+    normalization = (state.conj().T.dot(state)).todense()
```
Comment (mpharrigan) [inline referenceqvm/stabilizer_utils.py]: `.toarray()`

File: referenceqvm/stabilizer_utils.py [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+    Operators anticommute if symplectic inner product of their binary form is one
+
+    :param vector1: binary form of operator with no sign info
+    :param vector2: binary form of a pauli operator with no sign info
+    :return: 0, 1
+    """
+    if vector1.shape != vector2.shape:
+        raise ValueError("vectors must be the same size.")
+
+    # TODO: add a check for binary or integer linear arrays
```
Comment (mpharrigan) [inline referenceqvm/stabilizer_utils.py:223]: do you want to do this?

File: referenceqvm/stabilizer_utils.py [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+
+    hadamard_product = np.multiply(vector1, vector2)
```
Comment (mpharrigan) [inline referenceqvm/stabilizer_utils.py:225]: why not `vector1 * vector2`

File: referenceqvm/qam.py (hunk 4 above)
Comment (ncrubin) [reply referenceqvm/qam.py:96]: yes, but that was always a risk of defgates that was weighed in the original quil spec.  If users want to use defgate they should know that it takes them outside of the QAM model.  Remember QAM requires gate specification in the definition.  If the user wants to add something to this list then we are trusting them to know what they're doing.

File: referenceqvm/qvm_stabilizer.py (hunk 5 above)
Comment (ncrubin) [reply referenceqvm/qvm_stabilizer.py:87]: There is a better way to do this in terms of ABC's but that will be a future upgrade.  I will add to the issue list.

File: referenceqvm/qvm_stabilizer.py (hunk 8 above)
Comment (ncrubin) [reply referenceqvm/qvm_stabilizer.py:156]: Asymmetric Pauli channel model still works in stabilizer simulation.  preparing for the future

File: referenceqvm/qvm_stabilizer.py (hunk 9 above)
Comment (ncrubin) [reply referenceqvm/qvm_stabilizer.py:197]: moved to QAM superclass from wavefunction.  Fixes the same problem in density_qvm that slipped by!!!

File: referenceqvm/qvm_stabilizer.py (hunk 10 above)
Comment (ncrubin) [reply referenceqvm/qvm_stabilizer.py:189]: Future additions!  Right now this is just CHP simulator

File: referenceqvm/qvm_stabilizer.py (hunk 12 above)
Comment (ncrubin) [reply referenceqvm/qvm_stabilizer.py]: 👍

File: referenceqvm/qvm_stabilizer.py (hunk 11 above)
Comment (ncrubin) [reply referenceqvm/qvm_stabilizer.py:227]: 👍

File: referenceqvm/qvm_stabilizer.py (hunk 13 above)
Comment (ncrubin) [reply referenceqvm/qvm_stabilizer.py]: 👍

**Pull request merged**